### PR TITLE
RPC Notifier Signal when Setup Complete

### PR DESCRIPTION
**Pull request by brennanwatt**
#### Problem
See #27480. CI test flakiness

#### Summary of Changes
have `new_with_config` optionally signal back to callers when RPC Notifier thread pool routine has been installed. Have tests wait for setup to complete before proceeding with next steps.

Have `new_for_tests` call `new_for_tests_with_blockstore` to reduce redundant code

### CODE REVIEW

**Review by yihau — APPROVED** [review]
rpc_subscriptions::tests::test_check_account_subscribe has a same problem in my branch. After I patch this commit, it seems okay.

File: rpc/src/rpc_subscriptions.rs [hunk 1]
```diff
@@ -573,15 +573,18 @@ impl RpcSubscriptions {
...
             exit,
             max_complete_transaction_status_slot,
             blockstore,
             bank_forks,
             block_commitment_cache,
             optimistically_confirmed_bank,
             &PubSubConfig::default_for_tests(),
-        )
+        );
+        // Ensure RPC notifier is ready to receive notifications before proceeding
```
Comment (CriesofCarrots) [inline rpc/src/rpc_subscriptions.rs:585]: Hmm, this doesn't really ensure it's ready to receive notifications, does it? I'd prefer a deterministic check, if possible.

**Review by CriesofCarrots — APPROVED** [review]
Thanks for reworking this! lgtm if CI is happy

**Comment by brennanwatt** [comment]
> Thanks for reworking this! lgtm if CI is happy

Thanks for your help!

**Comment by yihau** [comment]
just noticed there is a similar error on v1.11 https://buildkite.com/solana-labs/solana/builds/81131#0182fea8-6a0c-47d9-8dd2-2f8eb55dc509
maybe we could backfill this PR to v1.10 and v1.11 🤔

**Comment by CriesofCarrots** [comment]
> maybe we could backfill this PR to v1.10 and v1.11 🤔

Since this is in essence a test-only change, backporting seems reasonable.

File: core/src/validator.rs [hunk 2]
```diff
@@ -722,6 +722,7 @@ impl Validator {
             block_commitment_cache.clone(),
             optimistically_confirmed_bank.clone(),
             &config.pubsub_config,
+            None,
```
Comment (hycinth22) [inline core/src/validator.rs:725]: 该处变更不重要。不关心ready信号则提供None值，关心则传入一个Option<Arc<AtomicBool>>

File: rpc/src/rpc_subscriptions.rs [hunk 3]
```diff
@@ -640,6 +656,9 @@ impl RpcSubscriptions {
                             .build()
                             .unwrap();
                         pool.install(|| {
+                            if let Some(rpc_notifier_ready) = rpc_notifier_ready {
+                                rpc_notifier_ready.fetch_or(true, Ordering::Relaxed);
```
Comment (hycinth22) [inline rpc/src/rpc_subscriptions.rs:660]: fn fetch_or(&self, val: bool, order: Ordering) -> bool
Logical “or” with a boolean value.

Performs a logical “or” operation on the current value and the argument val, and sets the new value to the result.

Returns the previous value.

File: rpc/src/rpc_subscriptions.rs [hunk 4]
```diff
@@ -640,6 +656,9 @@ impl RpcSubscriptions {
                             .build()
                             .unwrap();
                         pool.install(|| {
+                            if let Some(rpc_notifier_ready) = rpc_notifier_ready {
```
Comment (hycinth22) [inline rpc/src/rpc_subscriptions.rs:659]: 创建了一个线程池（notification_threads个线程）来处理process_notifications。 第一个启动的线程将设置rpc_notifier_ready为true

**Pull request merged**